apps/vaporgui/NavigationUtils.cpp

```diff
@@ -0,0 +1,403 @@
+    for (int i = 0; i < 3; i++) {
+        dirvec[i] = dirvec[i] * viewDist;
+        curPosVec[i] = (curCenter[i] - dirvec[i]);
+    }
+
+    SetAllCameras(ce, curPosVec, curViewDir, curUpVec, curCenter);
+}
+
+
+void NavigationUtils::SetAllCameras(ControlExec *ce, const double position[3], const double direction[3], const double up[3], const double origin[3])
+{
+    Trackball trackball;
+    bool      ok = trackball.setFromFrame(position, direction, up, origin, true);
+    if (!ok) {
+        MSG_ERR("Invalid camera settings");
+        return;
+    }
+
+    trackball.TrackballSetMatrix();
+    const double *m = trackball.GetModelViewMatrix();
+    SetAllCameras(ce, m, origin);
+}
+
+
+void NavigationUtils::SetAllCameras(ControlExec *ce, const vector<double> &position, const vector<double> &direction, const vector<double> &up, const vector<double> &origin)
+{
+    SetAllCameras(ce, position.data(), direction.data(), up.data(), origin.data());
+}
+
+
+void NavigationUtils::SetAllCameras(ControlExec *ce, const double matrix[16], const double origin[3])
+{
+    ParamsMgr *pm = ce->GetParamsMgr();
+    auto       vizNames = pm->GetVisualizerNames();
+    pm->BeginSaveStateGroup("Set Camera");
+    for (auto &viz : vizNames) {
+        ViewpointParams *vp = pm->GetViewpointParams(viz);
+        vp->SetModelViewMatrix(matrix);
+        vp->SetRotationCenter(origin);
+    }
+    pm->EndSaveStateGroup();
+}
+
+
+void NavigationUtils::SetAllCameras(ControlExec *ce, const vector<double> &m, const vector<double> &origin)
+{
+    assert(m.size() == 16);
+    SetAllCameras(ce, m.data(), origin.data());
+}
+
+
+void NavigationUtils::SetTimestep(ControlExec *ce, size_t ts)
+{
+    DataStatus *dataStatus = ce->GetDataStatus();
+    ParamsMgr * paramsMgr = ce->GetParamsMgr();
+
+    paramsMgr->BeginSaveStateGroup("Set Timestep");
+
+    // First set current *global* timestep in AnimationParams
+    //
+    AnimationParams *aParams = GetAnimationParams(ce);
+    aParams->SetCurrentTimestep(ts);
+
+    // Now set *local* time step for each RenderParams instance
+    //
+    vector<string> winNames = paramsMgr->GetVisualizerNames();
+    for (int i = 0; i < winNames.size(); i++) {
+        vector<string> dataSetNames = dataStatus->GetDataMgrNames();
+        for (int j = 0; j < dataSetNames.size(); j++) {
+            vector<RenderParams *> rParams;
+            paramsMgr->GetRenderParams(winNames[i], dataSetNames[j], rParams);
+
+            if (!rParams.size()) continue;
+
+            size_t local_ts = dataStatus->MapGlobalToLocalTimeStep(dataSetNames[j], ts);
+
+            for (int k = 0; k < rParams.size(); k++) {
+                RenderParams *rp = rParams[k];
+                rp->SetCurrentTimestep(local_ts);
+
+                // If variable was not initialzed but now is, reset the TF mapping range
+                std::string varNames[2] = {rp->GetVariableName(), rp->GetColorMapVariableName()};
```

Comment:
Should this be using rp->GetActualColorMapVariableName()?

Reply:
In this case it wouldn't really matter but I changed it to `GetActualColorMapVariableName`. Also, I did not write any of this code, it is legacy code that I moved around to clean up other parts of the GUI.